aips/aip-103.md

```diff
@@ -0,0 +1,206 @@
+
+Thus if we can leverage the existing design, it would be preferable to do so.
+
+## Specification and Implementation Details
+
+This AIP's implementation has three parts -
+
+1. We add an additional private input, `skip_aud_check`, into the prover.  This value will indicate whether the `aud` check is enabled.  
+- If it is disabled, the circuit will do the status quo set of verifications.  
+- If it is enabled, the circuit will use an empty string for the `aud` private input (as provided by the prover), which will be used as the `aud` value committed in the IdCommitment.  The circuit will skip matching the value of the JWT's aud claim with the `aud` private input (which is the empty string).  
+
+   Since the `aud` value is committed to in the `IdCommitment`, a proof generated with `skip_aud_check` will be rejected by the validator if the account is not `aud`-less, as the `IdCommitment` will be different, resulting in proof verification failing due to the public inputs hash being incorrect.
+
+   Similarly, a proof generated without `skip_aud_check` will be rejected by the validator if the account *is* `aud`-less, as in this case as well the `IdCommitment` will be different, resulting in proof verification failing due to the public inputs hash being incorrect.
+
+2. The prover API will also require an update to allow for indiciating whether the `aud` check is enabled.  This will be done by adding a new boolean argument `skip_aud_check` to the `prove` API.
+   
+   ```rust
+   #[derive(Debug, Serialize, Deserialize)]
+   pub struct RequestInput {
+      pub jwt_b64: String,
+      pub epk: EphemeralPublicKey,
+      #[serde(with = "hex")]
+      pub epk_blinder: EphemeralPublicKeyBlinder,
+      pub exp_date_secs: u64,
+      pub exp_horizon_secs: u64,
+      pub pepper: Pepper,
+      pub uid_key: String,
+      pub extra_field: Option<String>,
+      pub aud_override: Option<String>,
+      pub skip_aud_check: bool, // New argument
+   }
+   ```
+
+3. The pepper API will also require an update to allow for indiciating whether the `aud` check is enabled.  This is because for the pepper is derived using the `aud` value as one of the inputs, and for audless accounts the `aud` value needs to be the empty string in order for the account to be used across applications (which will differ by the value of the `aud` claim in the JWT).  Thus the pepper API needs to know whether the account is `aud`-less to construct the pepper appropriately.  This will be done by adding a new boolean argument `skip_aud_check` to the `fetch_pepper` API.
+
+   ```rust
+   #[derive(Debug, Deserialize, Serialize)]
+   pub struct PepperRequest {
+      pub jwt: String,
+      pub epk: EphemeralPublicKey,
+      pub exp_date_secs: u64,
+      pub epk_blinder: Vec<u8>,
+      pub uid_key: Option<String>,
+      pub derivation_path: Option<String>,
+      pub skip_aud_check: bool, // New argument
```

Comment:
I think the pepper service needs to limit uses of this parameter.

Specifically, it should probably only be allowed to be `true` for OIDC providers that we know are audless.

Reply:
Can you add this discussion somewhere, if it isn't there?